src/base64.cpp

```diff
@@ -14,36 +14,152 @@
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255};
 
-namespace cartesi {
+// Acumulates bytes in input buffer until 3 bytes are available.
+// Translate the 3 bytes into Base64 form and append to buffer.
+// Returns new number of bytes in buffer.
+static size_t b64encode(uint8_t c, uint8_t *input, size_t size, std::ostringstream &sout) {
+    input[size++] = c;
+    if (size == 3) {
+        uint8_t code[4];
+        unsigned long value = 0;
+        value += input[0];
+        value <<= 8;
+        value += input[1];
+        value <<= 8;
+        value += input[2];
+        code[3] = b64base[value & 0x3f];
+        value >>= 6;
+        code[2] = b64base[value & 0x3f];
+        value >>= 6;
+        code[1] = b64base[value & 0x3f];
+        value >>= 6;
+        code[0] = b64base[value];
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        sout << std::string_view(reinterpret_cast<char *>(code), 4);
+        size = 0;
+    }
+    return size;
+}
+
+// Encodes the Base64 last 1 or 2 bytes and adds padding '='
+// Result, if any, is appended to buffer.
+// Returns 0.
+static size_t b64pad(const uint8_t *input, size_t size, std::ostringstream &sout) {
+    unsigned long value = 0;
+    uint8_t code[4] = {'=', '=', '=', '='};
+    switch (size) {
+        case 1:
+            value = input[0] << 4;
+            code[1] = b64base[value & 0x3f];
+            value >>= 6;
+            code[0] = b64base[value];
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            sout << std::string_view(reinterpret_cast<char *>(code), 4);
+            break;
+        case 2:
+            value = input[0];
+            value <<= 8;
+            value |= input[1];
+            value <<= 2;
+            code[2] = b64base[value & 0x3f];
+            value >>= 6;
+            code[1] = b64base[value & 0x3f];
+            value >>= 6;
+            code[0] = b64base[value];
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            sout << std::string_view(reinterpret_cast<char *>(code), 4);
+            break;
+        default:
+            break;
+    }
+    return 0;
+}
+
+// Acumulates bytes in input buffer until 4 bytes are available.
+// Translate the 4 bytes from Base64 form and append to buffer.
+// Returns new number of bytes in buffer.
+static size_t b64decode(uint8_t c, uint8_t *input, size_t size, std::ostringstream &sout) {
+    // ignore invalid characters
+    if (b64unbase[c] > 64) {
```

Comment:
I think we should skip whitespaces but  throw an error otherwise.

Reply:
Ok, I explicitly defined ` \t\n\v\f\r` as the valid set for whitespaces.

Reply:
Since you went further, what do you think we should be implementing? The [MIME]( https://datatracker.ietf.org/doc/html/rfc2045) spec or the later [RFC](https://datatracker.ietf.org/doc/html/rfc4648). LuaSocket was used with MIME. But here we are not using MIME. In a nutshell, MIME requires encoding to split at most into 76 columns and asks decoding to ignore invalid characters. The later RFC says not to split into lines and to reject any invalid input. I guess we could stop breaking into lines when encoding but accept lines when decoding.

Reply:
I would say to not break lines when encoding, to save bytes and to make easier for others to decode base64 generated from us. If you agree, I will make the change.